### PR TITLE
code cleanup: Use libvmi in performance and replicaset tests

### PR DESCRIPTION
**Pull request by akrejcir**
**What this PR does / why we need it**:
Using libvmi instead of `NewRandomVMI` functions.

**Release note**:
```release-note
None
```

### CODE REVIEW

File: tests/performance/density.go [hunk 1]
```diff
@@ -156,8 +155,8 @@ func runAudit(startTime time.Time, endTime time.Time) {
 // createBatchVMIWithRateControl creates a batch of vms concurrently, uses one goroutine for each creation.
 // between creations there is an interval for throughput control
 func createBatchVMIWithRateControl(virtClient kubecli.KubevirtClient, vmCount int) {
-	for i := 1; i <= vmCount; i++ {
-		vmi := createVMISpecWithResources(virtClient)
+	for i := 0; i < vmCount; i++ {
```
Comment (EdDev) [inline tests/performance/density.go:158]: This is a hidden change in a commit that says that it changes something else.
Please do this in a separate commit or PR. Please also reason for it when you do.

File: tests/performance/realtime.go [hunk 2]
```diff
@@ -42,37 +39,39 @@ func byStartingTheVMI(vmi *v1.VirtualMachineInstance, virtClient kubecli.Kubevir
...
+		}
 	}
 }
 
 var _ = SIGDescribe("CPU latency tests for measuring realtime VMs performance", func() {
 
 	var (
-		vmi        *v1.VirtualMachineInstance
 		virtClient kubecli.KubevirtClient
-		err        error
```
Comment (EdDev) [inline tests/performance/realtime.go:71]: These changes are now mixed with the libvmi addition, but it would be easier to review and be confident that it is ok if you would have split them into separate commits, explaining why you did this.

**Review by EdDev — CHANGES_REQUESTED** [review]
Thanks.

Please add in the commit message the reason for this change (there are other PR/s proposing the same, you can copy it from there).
It would also be nice if you reduce the length of the subject line.
It is usually custom to make it 50 chars [1] and no longer than 72 [2].

[1] https://cbea.ms/git-commit/#limit-50
[2] https://cbea.ms/git-commit/#wrap-72

File: tests/performance/realtime.go [hunk 3]
```diff
@@ -80,8 +80,11 @@ var _ = SIGDescribe("CPU latency tests for measuring realtime VMs performance",
 	})
 
 	It("running cyclictest and collecting results directly from VM", func() {
-		vmi := tests.NewRandomVMIWithEphemeralDiskAndUserdata(cd.ContainerDiskFor(cd.ContainerDiskFedoraRealtime), tuneAdminRealtimeCloudInitData)
-		byConfiguringTheVMIForRealtime(vmi, "")
+		vmi := libvmi.NewFedora(
```
Comment (xpivarc) [inline tests/performance/realtime.go:83]: I don't think we should use `NewFedora` and I am not sure this work as you expect. Doesn't the two `libvmi.WithContainerImage` conflict?

File: tests/replicaset_test.go [hunk 4]
```diff
@@ -407,7 +406,7 @@ var _ = Describe("[rfe_id:588][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 
 	It("should replace a VMI immediately when a virt-launcher pod gets deleted", func() {
 		By("Creating new replica set")
-		template := tests.NewRandomVMIWithEphemeralDiskAndUserdata(cd.ContainerDiskFor(cd.ContainerDiskCirros), "#!/bin/bash\necho 'hello'\n")
+		template := libvmi.NewCirros()
```
Comment (xpivarc) [inline tests/replicaset_test.go:409]: Replacing `NewRandomVMIWithEphemeralDiskAndUserdata` with `NewCirros` is not direct imho.

File: tests/performance/density.go [hunk 5]
```diff
@@ -168,11 +167,9 @@ func createBatchVMIWithRateControl(virtClient kubecli.KubevirtClient, vmCount in
 }
 
 func createVMISpecWithResources() *kvv1.VirtualMachineInstance {
-	vmImage := cd.ContainerDiskFor("cirros")
 	const cpuLimit = "100m"
 	const memLimit = "90Mi"
-	cloudInitUserData := "#!/bin/bash\necho 'hello'\n"
-	vmi := tests.NewRandomVMIWithEphemeralDiskAndUserdata(vmImage, cloudInitUserData)
+	vmi := libvmi.NewCirros()
```
Comment (xpivarc) [inline tests/performance/density.go:172]: I am not sure this is really what we want. 
1. This doesn't use libvmi exclusively as we override memory below
2. We don't really follow the memory requirments here (128/256 Mi)
3.  We use different binding (not relevant to test as far as I can see)

**Review by xpivarc — COMMENTED** [review]
/hold
I believe the second commit is introducing bug.

File: tests/performance/density.go (hunk 5 above)
Comment (akrejcir) [reply tests/performance/density.go:172]: > 1. This doesn't use libvmi exclusively as we override memory below
I don't think this is a problem. There is no other way to set memory and cpu using libvmi.

> 2. We don't really follow the memory requirments here (128/256 Mi)
This point is irrelevant, because the memory is set below.

> 3. We use different binding (not relevant to test as far as I can see)
What binding do you mean? Can you explain?

File: tests/replicaset_test.go (hunk 4 above)
Comment (akrejcir) [reply tests/replicaset_test.go:409]: Ok. Do you know what is the difference? What parameters should I add?

Comment (xpivarc) [reply tests/replicaset_test.go:409]: Network binding is the difference. While it doesn't affect the test I would prefer to at least do this in 2 commits(One directly translates this to libvmi and in the second commit remove the binding)

File: tests/performance/density.go (hunk 5 above)
Comment (xpivarc) [reply tests/performance/density.go:173]: 1. I think we would benefit from defining the option. (limits can be a local option)
2. You are right
3.
 ```
vmi := libvmi.NewCirros(
					libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
					libvmi.WithNetwork(v1.DefaultPodNetwork()),
				)
```

Comment (akrejcir) [reply tests/performance/density.go]: done

File: tests/performance/realtime.go [hunk 6]
```diff
@@ -80,8 +80,13 @@ var _ = SIGDescribe("CPU latency tests for measuring realtime VMs performance",
 	})
 
 	It("running cyclictest and collecting results directly from VM", func() {
-		vmi = tests.NewRandomVMIWithEphemeralDiskAndUserdata(cd.ContainerDiskFor(cd.ContainerDiskFedoraRealtime), tuneAdminRealtimeCloudInitData)
-		byConfiguringTheVMIForRealtime(vmi, "")
+		vmi = libvmi.New(libvmi.RandName(),
+			libvmi.WithResourceMemory("512M"),
```
Comment (xpivarc) [inline tests/performance/realtime.go:84]: `libvmi.WithResourceMemory("512M")` is redundant, right?

**Review by xpivarc — COMMENTED** [review]
/approve
Thanks!

**Comment by akrejcir** [comment]
Can you also remove your `/hold` ?

File: tests/performance/realtime.go [hunk 7]
```diff
@@ -80,8 +80,13 @@ var _ = SIGDescribe("CPU latency tests for measuring realtime VMs performance",
 	})
 
 	It("running cyclictest and collecting results directly from VM", func() {
-		vmi = tests.NewRandomVMIWithEphemeralDiskAndUserdata(cd.ContainerDiskFor(cd.ContainerDiskFedoraRealtime), tuneAdminRealtimeCloudInitData)
-		byConfiguringTheVMIForRealtime(vmi, "")
+		vmi = libvmi.New(libvmi.RandName(),
```
Comment (xpivarc) [inline tests/performance/realtime.go:83]: `libvmi.RandName()`doesn't exist in main.

Comment (akrejcir) [reply tests/performance/realtime.go:83]: Thanks. 
Strange that it compiles on my machine with latest `main`.

Comment (akrejcir) [reply tests/performance/realtime.go:83]: Nevermind, It's probably because `make` does not compile functests.

Comment (xpivarc) [reply tests/performance/realtime.go]: Yes, you need `make functests`

**Comment by akrejcir** [comment]
@xpivarc or @EdDev, can you give `/lgtm`? Then this will be finally merged.

File: tests/performance/density.go [hunk 8]
```diff
@@ -196,3 +192,21 @@ func waitRunningVMI(virtClient kubecli.KubevirtClient, vmiCount int, timeout tim
...
+}
+
+func withCpuLimit(value string) libvmi.Option {
+	return func(vmi *kvv1.VirtualMachineInstance) {
+		if vmi.Spec.Domain.Resources.Limits == nil {
+			vmi.Spec.Domain.Resources.Limits = k8sv1.ResourceList{}
+		}
+		vmi.Spec.Domain.Resources.Limits[k8sv1.ResourceCPU] = resource.MustParse(value)
+	}
+}
```
Comment (EdDev) [inline tests/performance/density.go:212]: These two seem to be very basic and without any special logic. I think they can be placed under `libvmi`.

**Review by EdDev — APPROVED** [review]
Looks great, thank you!

I had one inline comment, to move two functions to `libvmi`. But this is not a must.

If you prefer not to do it, remove the hold.
/hold

File: tests/performance/density.go (hunk 8 above)
Comment (akrejcir) [reply tests/performance/density.go]: done

File: tests/performance/realtime.go [hunk 9]
```diff
@@ -43,24 +40,27 @@ func byStartingTheVMI(vmi *v1.VirtualMachineInstance, virtClient kubecli.Kubevir
...
-	vmi.Spec.Domain.CPU = &v1.CPU{
-		Model:                 "host-passthrough",
-		DedicatedCPUPlacement: true,
-		Realtime:              &v1.Realtime{Mask: realtimeMask},
-		NUMA:                  &v1.NUMA{GuestMappingPassthrough: &v1.NUMAGuestMappingPassthrough{}},
-	}
-	vmi.Spec.Domain.Memory = &v1.Memory{
-		Hugepages: &v1.Hugepages{PageSize: "2Mi"},
-		Guest:     &memoryRequest,
+func withRelatimeConfiguration(realtimeMask string) libvmi.Option {
```
Comment (EdDev) [inline tests/performance/realtime.go:43]: I think `realtimeMask` is not used (even before), can you add a commit before this one that removes it?

File: tests/performance/realtime.go [hunk 10]
```diff
@@ -43,24 +40,27 @@ func byStartingTheVMI(vmi *v1.VirtualMachineInstance, virtClient kubecli.Kubevir
...
+	memoryRequest := resource.MustParse("512Mi")
+	return func(vmi *v1.VirtualMachineInstance) {
+		vmi.Spec.Domain.Resources.Limits = k8sv1.ResourceList{
+			k8sv1.ResourceMemory: memoryRequest,
+			k8sv1.ResourceCPU:    resource.MustParse("2"),
+		}
+		vmi.Spec.Domain.Resources.Requests = k8sv1.ResourceList{
+			k8sv1.ResourceMemory: memoryRequest,
+			k8sv1.ResourceCPU:    resource.MustParse("2"),
+		}
```
Comment (EdDev) [inline tests/performance/realtime.go:53]: Can you use now `libvmi.WithResource*` and `libvmi.WithLimit*`? I think for the memory and CPU these exists now.

File: tests/libvmi/vmi.go [hunk 11]
```diff
@@ -88,9 +88,40 @@ func WithRng() Option {
...
-			k8sv1.ResourceMemory: resource.MustParse(value),
+		if vmi.Spec.Domain.Resources.Requests == nil {
+			vmi.Spec.Domain.Resources.Requests = k8sv1.ResourceList{}
 		}
+		vmi.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory] = resource.MustParse(value)
+	}
+}
+
+// WithResourceCpu specifies the vmi CPU resource.
+func WithResourceCpu(value string) Option {
```
Comment (EdDev) [inline tests/libvmi/vmi.go:99]: We should probably create a new file named `resources` and place these there. The `vmi.go` is increasing too much.
If you have time, it will be nice to see a PR moving there the existing ones (can be a follow up PR).

**Review by EdDev — APPROVED** [review]
Thank you!
There is one comment I think it is worth attending to (`libvmi.WithResource*` and `libvmi.WithLimit*`), the rest do not have to come in the PR for sure (but they can).

Leaving it up to you to decide how to proceed. You can remove the hold I placed originally based on that.

File: tests/performance/realtime.go (hunk 10 above)
Comment (akrejcir) [reply tests/performance/realtime.go]: done

File: tests/performance/realtime.go (hunk 9 above)
Comment (akrejcir) [reply tests/performance/realtime.go]: I will to it in future PR

File: tests/libvmi/vmi.go (hunk 11 above)
Comment (akrejcir) [reply tests/libvmi/vmi.go:99]: Good idea, I will do it in a future PR.

**Comment by akrejcir** [comment]
> Please fix the [linter errors](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7552/pull-kubevirt-code-lint/1546773411429093376#1:build-log.txt%3A139).
> 
> You can run `make lint` locally to see it passes before you push.

Done. I'm used to camel case names also for acronyms like CPU or VM.

> Also, a small request to ease the review of changes, please do not combine a rebase with changes, try to push them separately so it can be easily reviewed (instead of going over all the changes from start).

Good idea, I will do it next time.

**Review by EdDev — APPROVED** [review]
Thanks!

**Review by EdDev — APPROVED** [review]
